buildSrc/plugins/src/main/kotlin/org/jetbrains/kotlin/Utils.kt

```diff
@@ -1,24 +1,115 @@
+            if (this is Array<*>) this.toList() as List<String>
+            else this as List<String>
+    }
+
+//endregion
 
 /**
- * Ad-hoc signing of the specified path
+ * Ad-hoc signing of the specified path.
  */
 fun codesign(project: Project, path: String) {
-    check(HostManager.hostIsMac, { "Apple specific code signing" } )
+    check(HostManager.hostIsMac) { "Apple specific code signing" }
     val (stdOut, stdErr, exitCode) = runProcess(executor = localExecutor(project), executable = "/usr/bin/codesign",
             args = listOf("--verbose", "-s", "-", path))
-    check(exitCode == 0, { """
-            |Codesign failed with exitCode: $exitCode
-            |stdout: $stdOut
-            |stderr: $stdErr
-            """.trimMargin()
-    })
-}
\ No newline at end of file
+    check(exitCode == 0) { """
+        |Codesign failed with exitCode: $exitCode
+        |stdout: $stdOut
+        |stderr: $stdErr
+        """.trimMargin()
+    }
+}
+
+/**
+ * Creates a list of file paths to be compiled from the given [compile] list with regard to [exclude] list.
+ */
+fun Project.getFilesToCompile(compile: List<String>, exclude: List<String>): List<String> {
+    // convert exclude list to paths
+    val excludeFiles = exclude.map { project.file(it).absolutePath }.toList()
+
+    // create list of tests to compile
+    return compile.flatMap { f ->
+        project.file(f)
+                .walk()
+                .filter { it.isFile && it.name.endsWith(".kt") && !excludeFiles.contains(it.absolutePath) }
+                .map(File::getAbsolutePath)
+                .asIterable()
+    }
+}
+
+//region Task dependency.
+
+fun Project.setDistDependencyFor(taskName: String) {
+    project.setDistDependencyFor(project.tasks.getByName(taskName))
+}
+
+fun Project.setDistDependencyFor(t: Task) {
+    val rootTasks = project.rootProject.tasks
+    // We don't build the compiler if a custom dist path is specified.
+    if (!(project.findProperty("useCustomDist") as Boolean)) {
+        t.dependsOn(rootTasks.getByName("dist"))
+        val target = project.testTarget
+        if (target != HostManager.host) {
+            // if a test_target property is set then tests should depend on a crossDist
+            // otherwise runtime components would not be build for a target.
+            t.dependsOn(rootTasks.getByName("${target.name}CrossDist"))
+        }
+    }
+}
+
+/**
+ * @see setDistDependencyFor
+ */
+fun String.setSameDependenciesAs(task: Task) {
+    val t = task.project.tasks.getByName(this)
+    t.setSameDependenciesAs(task)
+}
+
+/**
+ * Sets the same dependencies for the receiver task from the given [task]
+ * and make [task] depend on receiver
+ */
+fun Task.setSameDependenciesAs(task: Task) {
+    val dependencies = task.dependsOn.toList() // save to the list, otherwise it will cause cyclic dependency.
+    this.dependsOn(dependencies)
+    // Run task should depend on compile task.
+    task.dependsOn(this)
```

Comment:
This effect (making `task` depending on the receiver) doesn't look intuitive. May be remove it from this function and specify such a dependency manually at a caller side? Or at least rename the function to reflect this effect?